src/org/traccar/protocol/Arnavi4ProtocolDecoder.java

```diff
@@ -0,0 +1,187 @@
+        super(protocol);
+    }
+
+    private Position decodePosition(DeviceSession deviceSession, ChannelBuffer buf, int length, Date time) {
+
+        final Position position = new Position();
+        position.setProtocol(getProtocolName());
+        position.setDeviceId(deviceSession.getDeviceId());
+
+        position.setTime(time);
+
+        int readBytes = 0;
+        while (readBytes < length) {
+            short tag = buf.readUnsignedByte();
+            switch (tag) {
+                case TAG_LATITUDE:
+                    position.setLatitude(buf.readFloat());
+                    position.setValid(true);
+                    break;
+
+                case TAG_LONGITUDE:
+                    position.setLongitude(buf.readFloat());
+                    position.setValid(true);
+                    break;
+
+                case TAG_COORD_PARAMS:
+                    position.setCourse(buf.readUnsignedByte() * 2.0);
+                    position.setAltitude(buf.readUnsignedByte() * 10.0);
+                    byte satellites = buf.readByte();
+                    position.set(Position.KEY_SATELLITES, satellites & 0x0F + (satellites >> 4) & 0x0F); // gps+glonass
+                    position.setSpeed(buf.readUnsignedByte() * 1.852);
+                    break;
+
+                default:
+                    buf.readBytes(4); // Skip other tags
+                    break;
+            }
+
+            readBytes += 5; // 1 byte tag + 4 bytes value
+        }
+
+        return position;
+    }
+
+    @Override
+    protected Object decode(
+            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        ChannelBuffer buf = (ChannelBuffer) msg;
+
+        byte startSign = buf.readByte();
+
+        if (startSign == HEADER_START_SIGN) {
+
+            byte version = buf.readByte();
+
+            String imei = String.valueOf(buf.readLong());
+            DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, imei);
+
+            if (deviceSession != null && channel != null) {
+
+                final ChannelBuffer response;
+
+                if (version == HEADER_VERSION_1) {
+                    response = ChannelBuffers.dynamicBuffer(ByteOrder.LITTLE_ENDIAN, 4);
+                    response.writeBytes(new byte[]{0x7B, 0x00, 0x00, 0x7D});
+
+                } else if (version == HEADER_VERSION_2) {
+                    response = ChannelBuffers.dynamicBuffer(ByteOrder.LITTLE_ENDIAN, 9);
+                    response.writeBytes(new byte[]{0x7B, 0x04, 0x00});
+                    byte[] timeBytes = ByteBuffer.allocate(4).putInt((int) (System.currentTimeMillis() / 1000)).array();
+                    response.writeByte(Checksum.modulo256(timeBytes));
+                    response.writeBytes(timeBytes);
+                    response.writeByte(0x7D);
+
+                } else {
+                    throw new IllegalArgumentException("unsupported header version");
```

Comment:
I think this is unnecessary. We usually avoid throwing exceptions from decoder. This way there is a good chance that server would still work even if protocol is updated.